### PR TITLE
[Feature]: Added TensorDictDefaultInitializer

### PR DESCRIPTION
**Pull request by nicolas-dufour**
## Description

This PR adds a module that can initialize a set of tensors to default values and add it to the given TD
## Motivation and Context

This change is required because we might want to init some tensors inside of a TDSequence. For example if the td is given from an env, we can want to add some other tensors. This could require to make changes to core methods such as collectors, rollouts et... By doing this, we can directly do the change inside a TDSequence

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [x] New feature (non-breaking change which adds core functionality)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/facebookresearch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: test/test_tensordictmodules.py [hunk 1]
```diff
@@ -1494,7 +1496,7 @@ def test_vmap_probabilistic(self, safe, spec_type):
                 out_key_sample=["out"],
                 dist_param_keys=["loc", "scale"],
                 safe=safe,
-                **kwargs
+                **kwargs,
```
Comment (vmoens) [inline test/test_tensordictmodules.py:1499]: why are all these changes necessary?

File: test/test_tensordictmodules.py [hunk 2]
```diff
@@ -1701,6 +1703,100 @@ def test_subsequence_weight_update(self):
...
+    @pytest.mark.parametrize("batch_size", [[], [3], [5], [2, 3]])
+    @pytest.mark.parametrize("initializer", [torch.zeros, torch.ones, torch.randn])
+    @pytest.mark.parametrize("num_new_tensors", [1, 2, 3])
+    def test_td_initializer(self, device, batch_size, initializer, num_new_tensors):
+        td = TensorDict(
+            {"a": torch.randn(*batch_size, 3), "b": torch.randn(*batch_size, 4)},
+            batch_size,
+        ).to(device)
+        td_ref = td.clone()
+        default_td = {
```
Comment (vmoens) [inline test/test_tensordictmodules.py:1718]: why is it named `*_td`?

File: test/test_tensordictmodules.py [hunk 3]
```diff
@@ -1701,6 +1703,100 @@ def test_subsequence_weight_update(self):
...
+            f"c_{i}": {"initializer": initializer, "shape": [2]}
+            for i in range(num_new_tensors)
+        }
+        initializer = TensorDictDefaultInitializer(default_td).to(device)
+        td = initializer(td)
+
+        assert torch.allclose(td["a"], td_ref["a"])
+        assert torch.allclose(td["b"], td_ref["b"])
+        for i in range(num_new_tensors):
+            assert f"c_{i}" in td.keys()
```
Comment (vmoens) [inline test/test_tensordictmodules.py:1728]: Ideally I guess you'd like to test the value

File: test/test_tensordictmodules.py [hunk 4]
```diff
@@ -1701,6 +1703,100 @@ def test_subsequence_weight_update(self):
...
+            assert td[f"c_{i}"].shape == torch.Size([*batch_size, 2])
+            assert td[f"c_{i}"].device == device
+
+    @pytest.mark.parametrize("batch_size", [[], [3], [5], [2, 3]])
+    @pytest.mark.parametrize("initializer", [torch.zeros, torch.ones, torch.randn])
+    @pytest.mark.parametrize("num_new_tensors", [1, 2, 3])
+    def test_td_initializer_no_device(self, batch_size, initializer, num_new_tensors):
```
Comment (vmoens) [inline test/test_tensordictmodules.py:1735]: why do we need this? What info does it bring wrt the previous one?

File: torchrl/modules/tensordict_module/initializer.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        params: Optional[Union[TensorDictBase, List[Tensor]]] = None,
+        buffers: Optional[Union[TensorDictBase, List[Tensor]]] = None,
+        **kwargs,
+    ) -> TensorDictBase:
+        batch_size = tensordict.batch_size
+        device = tensordict.device_safe()
+        tensors = self.module(self.defaults_init, batch_size, device)
+        if not self.reinit:
+            tensors = {k: v for k, v in tensors.items() if k not in tensordict}
+        new_tensors_td = TensorDict(tensors, batch_size=batch_size)
```
Comment (vmoens) [inline torchrl/modules/tensordict_module/initializer.py:108]: creating a TD is more expensive than a dict, perhaps we can just work with a dict?

**Review by vmoens — COMMENTED** [review]
I see the utility in the case of Dreamer, but is there any other module that can have no input and has to figure out what to do?
I feel that this is a bit overkilling it, and if there are no other usages on the horizon I would put it on hold. I feel that users can create a non-generic TensorDictModule that does this if they need to.
TensorDictModule is restrictive but I guess it's fine: it can't read strings, its inputs should only be tensors etc. We should probably figure out how to make this more modular but I guess we also want to minimize the number of primitives we're supporting in case there's a need to refactor the whole class.

In the case that is ours (Dreamer) I feel that the problem is more of a design question:
the policy depends on the action at t-1, so even for the first action we need to have a 0th action available. I think a Transform could do that. We could do a transform that creates a tensor of arbitrary key, size, dtype and with a certain value that will be placed on the TensorDict when reset. There are multiple use cases for this: setting a 0th action, setting a 0th hidden state, setting an env-specific indicator (e.g. the task in multi-task settings, or the agent in multi-agent)

I'm not saying we should close this but I'd rather put it on hold until we have a stronger signal that this is needed.

Happy to hear your thoughts about it

**Comment by nicolas-dufour** [comment]
Closing this, can be reopen if a need for this kind of tensordictmodule primitive arises

**Pull request closed without merging**